### PR TITLE
[GPII-3195]: Add affinity setting to CouchDB

### PR DESCRIPTION
**Pull request by natarajaya**
This solves the problem perfectly.
Chart changes should go to the https://github.com/helm/charts I guess?

### CODE REVIEW

**Comment by stepanstipl** [comment]
Looks good. Do we want `preferredDuringSchedulingIgnoredDuringExecution` or `required...`? The latter one prevents pods from running if there are not enough nodes in a cluster.

**Comment by mrtyler** [comment]
LGTM

> Do we want preferredDuringSchedulingIgnoredDuringExecution or required...?

On one hand, `preferred` may give us a little more resiliency since it is "easier" to find a Node where a Pod can run. On the other hand, I think the `required` behavior is more straightforward / less confusing since a Pod either runs in the "right" place or not at all.

In practice, the requirement on PVCs (which are stuck to a particular AZ on both AWS and GCP) has produced the `required` behavior (couchdb-2 needs the PVC in AZ 2, so it can only run on the Node in AZ 2).

I don't have a strong opinion, but I lean toward `required`. Am I missing anything about the trade-offs?

File: common/charts/couchdb/values.yaml [hunk 1]
```diff
@@ -54,6 +54,12 @@ helperImage:
 ## `OrderedReady`
 podManagementPolicy: Parallel
 
+
+## To better tolerate Node failures, we can enable Kubernetes scheduler to not assign
+## more than one Pod of CouchDB StatefulSet per Node using podAntiAffinity.
+## This is still a beta feature, and will get stable in Kubernetes 1.2
```
Comment (seg10) [inline common/charts/couchdb/values.yaml]: `Kubernetes 1.2` is probably the wrong version! 1.2 looks like the version that affinity was introduced. Definitely a case of the documentation lying to the reader.

File: common/charts/couchdb/templates/statefulset.yaml [hunk 2]
```diff
@@ -21,6 +21,18 @@ spec:
         app: {{ template "couchdb.name" . }}
         release: {{ .Release.Name }}
     spec:
+    {{- if .Values.podAntiAffinity }}
```
Comment (seg10) [inline common/charts/couchdb/templates/statefulset.yaml]: Given how expressive the label selector that gets expanded is, it seems like `.Values.podAntiAffinity` doesn't really communicate the intent for this code.  I recommend finding another term. Some suggestions come to mind:

1. .Values.spreadPods
2. .Values.spreadPodsUniformly
3. .Values.distributePods

File: common/charts/couchdb/values.yaml (hunk 1 above)
Comment (natarajaya) [reply common/charts/couchdb/values.yaml]: Nice catch!

File: common/charts/couchdb/templates/statefulset.yaml (hunk 2 above)
Comment (natarajaya) [reply common/charts/couchdb/templates/statefulset.yaml]: Yeah, this seemed too confusing to me as well. I just made `affinity` fully configurable with one example of `podAntiAffinity` in default chart values.

Comment (seg10) [reply common/charts/couchdb/templates/statefulset.yaml]: Nice! I think that will work quite well.

**Comment by natarajaya** [comment]
@amatas @mrtyler @stepanstipl @seg10 Thanks for your reviews!

Instead of somewhat confusing true/false switch on `podAntiAffinity`, I made `affinity` parameter fully configurable for the chart.

Regarding arguments about particular affinity settings, I would prefer to stick to `requiredDuringSchedulingIgnoredDuringExecution` rather than `preferred`, we don't want to have more than 1 CouchDB replica per node in almost any "normal" circumstances anyway, due to current CouchDB replication configuration. It also seems like after upgrade to K8s 1.9.7, `preferred` behavior is now default one.

Adding `failure-domain.beta.kubernetes.io/zone` apart from a hostname will prevent us from having more DB replicas than number of the cluster zones, which is not ideal. Since affinity is fully configurable, we can add more conditions in there later if needed (maybe we'll have specific instance flavor for DB nodes, so we need to make DB pods stick to nodes with some additional label, etc.)

We do not manage PVCs directly anymore, but yes, using custom StorageClass can add a little more flexibility to this, if needed.

I think this is ready to be merged now, I'll also prepare changes for upstream charts.

**Comment by mrtyler** [comment]
LGTM

Thanks for laying out your rationale.

**Comment by natarajaya** [comment]
Created https://github.com/helm/charts/pull/7224

**Pull request merged**